### DISCUSSION TITLE
Who is the owner of code generated by capnpc?

### DISCUSSION THREAD

**Discussion by ghost** [capnproto/capnproto#2162] [Q&A] [answered]
Hi, thank you for the project.

Could you please let me clarify the owner of code generated  by capnpc?

For example, in protobuf case, it's owned by the owner of the input file used when generating it, 
according to [protobuf LICENSE](https://github.com/protocolbuffers/protobuf/blob/10f611a1b5f2a4ffe9ab380ce1aee04f78d498cb/LICENSE#L29-L32)

> Code generated by the Protocol Buffer compiler is owned by the owner
of the input file used when generating it.  This code is not
standalone and requires a support library to be linked with it.  This
support library is itself covered by the above license.

Thank you!

**Comment by kentonv** [comment, marked as answer]
The output of the Cap'n Proto compiler is a mixture of things derived from its input schema file and things derived from Cap'n Proto's own source code. I am not a lawyer, but I believe this means that it would be considered a derivative work of both. This would make it subject to Cap'n Proto's license (the MIT license) as well as any license on the input file.

Since the generated code is dependent on the Cap'n Proto library anyway, this is not meaningfully different from if we considered the output to be wholly owned by the owner of the input -- either way, your final program is going to be a combined derivative work.

Since the MIT license is permissive, I'd expect this doesn't present any actual legal problems to anyone. That said, if there is actually a reason adding a note like Protobuf's would make a difference, I am not against adding it.

------

Fun fact: The note in the Protobuf license was actually added by me, 16 years ago:

https://github.com/protocolbuffers/protobuf/commit/c7f06b2ad5b030223d1408a5ef8fea464b64f998

But I don't think I wrote it myself, I asked a Google lawyer to write it. I don't have such a lawyer to review license changes for Cap'n Proto, which makes me a little more hesitant to use anything other than the straight MIT license, but I could be convinced. Personally I suspect this note isn't actually legally meaningful, we only added it because someone asked.

**Reply by ghost** [reply]
@kentonv 

Thank you very much for your quick reply!
Your answer makes sense for me.
Therefore I don't ask you to modify license.

> Since the generated code is dependent on the Cap'n Proto library anyway, this is not meaningfully different from if we considered the output to be wholly owned by the owner of the input -- either way, your final program is going to be a combined derivative work.

If you change capnproto license to another than MIT,  there might be meaningful difference
because some project can statically link generated code and dynamically link capnproto library.
So please consider to modify license if you change capnproto license itself.

> Fun fact: The note in the Protobuf license was actually added by me, 16 years ago:
>
> https://github.com/protocolbuffers/protobuf/commit/c7f06b2ad5b030223d1408a5ef8fea464b64f998

Amazing! :superhero:

**Reply by kentonv** [reply]
Hmm, I don't fully know how dynamic linking and open source licenses interact. But, certainly, if you distributed a dynamically-linked binary, and someone chose to use it together with a version of capnp under a different license, it would not be your responsibility -- it's up to that user to make sure they are complying with the new license.

Anyway, the license on Cap'n Proto is not likely to change... especially not in a patch release that you could actually dynamically link with binaries built against older versions. (Usually each version of Cap'n Proto is ABI-incompatible so dynamic linking doesn't actually work that well anyway...)

**Reply by ghost** [reply]
> Hmm, I don't fully know how dynamic linking and open source licenses interact. But, certainly, if you distributed a dynamically-linked binary, and someone chose to use it together with a version of capnp under a different license, it would not be your responsibility -- it's up to that user to make sure they are complying with the new license.

OK, let me clarify.
I'm concerning about the following scenario (Of course, this is just example).

1. CapnProto license becomes L-GPL
2. My application links CapnProto library dynamically
3. My application links generated code from CapnProto schema statically

If the license of generated code from CapnProto schema is L-GPL as CapnProto, my application must be L-GPL.
If the license of generated code from CapnProto schema can be decided by me, my application can select license as it is.

> Anyway, the license on Cap'n Proto is not likely to change... 

OK, got it!
Thanks!

**Reply by kentonv** [reply]
> If the license of generated code from CapnProto schema is L-GPL as CapnProto, my application must be L-GPL.

This would only apply if you yourself used an LGPL version of Cap'n Proto to build your application, and then distributed the binary. If you built your binary against an MIT-licensed version, and then, later on, a new version of the dynamic library is licensed LGPL, that license change does not retroactively apply to your binary that was built against the MIT version.

But in any case, I would never relicense Cap'n Proto with a GPL-ish license. :)

**Reply by ghost** [reply]
> But in any case, I would never relicense Cap'n Proto with a GPL-ish license. :)

Okay, I'm glad to hear that!
Thank you for the discussion!
